mlabns/mlabns.go

```diff
@@ -0,0 +1,126 @@
+// Package mlabns implements a simple mlab-ns client.
+package mlabns
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+// httpRequestor is the interface of the implementation that
+// performs a mlabns HTTP request for us.
+type httpRequestor interface {
+	// Do performs the request and returns either a response or
+	// a non-nil error to the caller.
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// httpRequestMaker is the type of the function that
+// creates a new HTTP request for us.
+type httpRequestMaker = func(
+	method, url string, body io.Reader) (*http.Request, error,
+)
+
+// Client is an mlabns client.
+type Client struct {
+	// BaseURL is the optional base URL for contacting mlabns. This is
+	// initialized in NewClient, but you may override it.
+	BaseURL string
+
+	// Ctx is the context to use.
+	Ctx context.Context
+
+	// Tool is the mandatory tool to use. This is initialize by NewClient.
+	Tool string
+
+	// UserAgent is the mandatory user agent to be used. Also this
+	// field is initialized by NewClient.
+	UserAgent string
+
+	// requestMaker is the function that creates a request. This is
+	// initialized in NewClient, but you may override it.
+	requestMaker httpRequestMaker
+
+	// requestor is the implementation that performs the request. This is
+	// initialized in NewClient, but you may override it.
+	requestor httpRequestor
+}
+
+// baseURL is the default base URL.
+//
+// TODO(bassosimone): when ndt7 is deployed on the whole platform, we can
+// stop using the staging mlabns service and use the production one.
+const baseURL = "https://locate-dot-mlab-staging.appspot.com/"
+
+// NewClient creates a new Client instance with mandatory userAgent, and tool
+// name. For running ndt7, use "ndt7" as the tool name.
+func NewClient(ctx context.Context, tool, userAgent string) *Client {
+	return &Client{
+		BaseURL:      baseURL,
+		Ctx:          ctx,
+		requestMaker: http.NewRequest,
+		requestor:    http.DefaultClient,
```

Comment:
Since I'm already wrapping the request with a context, I'll implement this by passing a sub-context that has the desired timeout. I think that "mlab-ns hasn't responded" should take into account not only the connection itself, but also receiving the response. So, I'm going for the sub-context road rather than limiting myself to a dialer.

BTW, this was a great comment. I have also implemented a timeout for the websocket handshake, using features that gorilla/websocket already has.

Also, when reasoning on this comment, I did some yak shaving which resulted in:

1. the `ndt7.Client` constructor takes a mandatory user-agent and the documentation says mlab may reject empty user agents in the future (so, we nudge people towards identifying them by the structure of the code and the documentation)

2. the `ndt7.Client` now has a `*mlabns.Client` parameter and `ndt7.NewClient` constructs such `mlabns.Client` with default-good arguments -- this change is enabled by the previous one and allows me to make the code less repetitive

(Will push soon.)